### ISSUE TITLE
How to reproduce Supervised Open Information Extraction？

### ISSUE THREAD

**Issue by piekey1994** [allenai/allennlp#2840]
How to reproduce Supervised Open Information Extraction
I want to reproduce Supervised Open Information Extraction. I used the model.py file at https://github.com/gabrielStanovsky/supervised-oie to train. But the results I got were lower than the 62% F1-score, only about 50%. I would like to know:
1. My 50% F1-score is obtained from sklearn.metrics.f1_score(Y, y, average = 'micro'). The F1-score in the paper does not seem to be calculated by this, right?
2. Is the implementation details of Open Information Extraction in AllenNLP different from https://github.com/gabrielStanovsky/supervised-oie?
Thanks a lot.

**Comment by schmmd** [comment]
@gabrielStanovsky are you able to answer this question?

**Comment by gabrielStanovsky** [comment]
Hi @piekey1994 ,
1. The F1 score in the paper is calculated using a confidence threshold which maximizes F1 on the dev set. Which confidence score threshold did you use to compute the F1 score you got?
2. The Open IE model in AllenNLP is a reimplementation of that in https://github.com/gabrielStanovsky/supervised-oie , which benefits from an easier and standardized way to do trainining and evaluation.

**Comment by piekey1994** [comment]
Thank you @gabrielStanovsky ,
1. I need to save the model during training, when it evaluates the dev set to get the maximum F1 score, right? Although I still don't understand what the confidence score threshold means. I didn't seem to use this indicator when calculating F1.

2. I found out from the config of AllenNLP's OpenIE model that it is similar to the structure of Deep SRL. This openie model is implemented by stacking alternating_lstm. I have seen your keras version for a long time https://github.com/gabrielStanovsky/supervised-oie. This code does not seem to fully implement the highway and RNN dropout in alternating_lstm. So I am trying to implement these features in the original keras version. (Maybe using AllenNLP directly would be a better idea ^_^)

**Comment by piekey1994** [comment]
Hi @gabrielStanovsky ,
I seem to understand the meaning of the confidence threshold you said. For the model generation probability, I calculate the P value and the R value based on this probability, instead of simply taking 0.5 as the threshold. I want to adjust this threshold when calculating the F1 value so that F1 can be maximized.

**Comment by piekey1994** [comment]
Hi @gabrielStanovsky ,
Is the f1 value in your paper calculated by this benchmark.py in https://github.com/gabrielStanovsky/oie-benchmark ? I read the source code and understood the meaning of the confidence threshold. However, during the training process, if the model generated for each epoch is used to calculate the F1 value, it is very troublesome.

**Comment by gabrielStanovsky** [comment]
Hi @piekey1994 ,
I'm not sure I understand the concern. 
I hope this helps to clarify things (see Section 4 in [our paper](https://gabrielstanovsky.github.io/assets/papers/naacl18long/paper.pdf) for more details):
- During training: The model is trained to minimize per-label categorical loss (BIO tags for each word).
- During inference: We calculate other metrics, such as precision and recall with respect to Open IE extractions, as well as the model's confidence in each extraction, calculated as the multiplication of the softmax's probability in each of the labels which make up the extraction. We can use this confidence approximation to filter extractions in which the model is less confident.
- Following, we can draw a precision-recall curve for each confidence threshold (Fig. 5), and also calibrate a threshold which maximizes F1 on the dev set to arrive at a single F1 score.

This doesn't mean that this is the "right" threshold, as different users may have different needs (e.g., maximize precision vs. maximizing recall).

**Comment by gabrielStanovsky** [comment]
In any case, perhaps this is not the best place for a discussion about a specific research paper which doesn't concern the framework itself? 
I'm happy to move this discussion to https://github.com/gabrielStanovsky/oie-benchmark or email if you prefer.

**Comment by gabrielStanovsky** [comment]
@piekey1994 Closing for now, as this seems to concern a research question regarding a specific model, rather than an issue with the library. 
I hope the information above was helpful, and please reach out if you want to continue this discussion.

**Comment by piekey1994** [comment]
Hi @gabrielStanovsky ,
Thank you very much for your detailed and effective help and advice.
